invisible_cities/io/dst_io.py

```diff
@@ -39,28 +39,43 @@ def _make_tabledef(column_types : pd.Series, str_col_length : int=32) -> dict:
             tabledef[colname] = tb.Col.from_type(coltype, pos=indx)
     return tabledef
 
-def _store_pandas_as_tables(h5out : tb.file.File, df : pd.DataFrame, group_name : str, table_name : str, compression : str='ZLIB4', descriptive_string : [str]="", str_col_length : int=32) -> None:
+def store_pandas_as_tables(h5out : tb.file.File, df : pd.DataFrame, group_name : str, table_name : str, compression : str='ZLIB4', descriptive_string : [str]="", str_col_length : int=32) -> None:
     if len(df) == 0:
         warnings.warn(f'dataframe is empty', UserWarning)
     if group_name not in h5out.root:
         group = h5out.create_group(h5out.root, group_name)
-
+    arr = df.to_records(index=False)
     group = getattr(h5out.root, group_name)
+
     if table_name not in group:
-        tabledef = _make_tabledef(df.dtypes)
+        tabledef = _make_tabledef(df.dtypes, str_col_length=str_col_length)
         table    =  make_table(h5out,
                                group       = group_name,
                                name        = table_name,
                                fformat     = tabledef,
                                description = descriptive_string,
                                compression = compression)
+        #hack to transform object numpy dtype to string
+        def _cast_type(dtype : np.dtype):
+            if dtype != 'O': return dtype
+            else:            return f'S{str_col_length}'
+        data_types = np.dtype([(col, _cast_type(arr[col].dtype)) for col in arr.dtype.names])
+    else:
+        table = getattr(group, table_name)
+        data_types = table.dtype
 
-    table = getattr(group, table_name)
     if not np.array_equal(df.columns, table.colnames):
         raise TableMismatch(f'dataframe differs from already existing table structure')
-    for indx in df.index:
-        tablerow = table.row
-        for colname in table.colnames:
-            tablerow[colname] = df.at[indx, colname]
-        tablerow.append()
+
+    #check which columns are of type object and throw warning if length exceeds
+    for colname in df.columns:
+        if (df[colname].dtype.name == 'object'):
```

Comment:
Maybe this could be written as
```
for colname, col in filter(lambda (_, c): c.dtype.name == "object", df.items()):
    if col.str.len().max() > data_types[colname].itemsize:
        warnings.warn...
```
or
```
for colname, col in df.items():
    if col.dtype.name == "object" and col.str.len().max() > data_types[colname].itemsize:
        warnings.warn...
```

Reply:
Right... Let me try to rewrite the whole function in a more sensible way...